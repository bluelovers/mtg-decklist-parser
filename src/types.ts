import { CardModel } from './cardModel';

export interface ICardWithoutAmount
{
	/**
	 * The amount of the card specified within the deck.
	 */
	name: string,
	/**
	 * The set code of the card.
	 */
	set?: string,
	/**
	 * The collector number of the card within the specific set.
	 */
	collectors?: number,
	/**
	 * The ID of the card specific to MTG Online.
	 */
	mtgoID?: string,
}

export interface ICard extends ICardWithoutAmount
{
	/**
	 * The name of the card specified within the deck.
	 */
	amount: number,
}

export interface ICardXmlObject
{
	Quantity: string,
	Name: string,
	CatID: string,
}

export interface IDeck
{
	/**
	 * deck name
	 */
	name?: string;

	/**
	 * An array of `CardModel` for the main deck.
	 */
	deck: CardModel[];
	/**
	 * An array of `CardModel` for the sideboard.
	 */
	sideboard: CardModel[];
	/**
	 * If a companion is specified in the input will be available, otherwise null.
	 */
	companion?: CardModel;
	/**
	 * If a commander is specified in the input will be available, otherwise null.
	 */
	commander?: CardModel;
}

export const SymDecklistType = Symbol.for(`DecklistType`)

export const enum EnumDecklistType
{
	mtgo = 'mtgo',
	mtga = 'mtga',
	mtgify = 'mtgify',
}
