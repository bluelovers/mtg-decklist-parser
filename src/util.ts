import { ICard, ICardWithoutAmount, IDeck } from './types';

//const _amountRegex = /^\d+/;
const _collectorRegex = /\d+$/;
const _setRegex = /(\(|\[)(.+)(\)|\])/;

export function parseString(rawInput: string): ICard
{
	rawInput = rawInput.trim();

	let m_amount = rawInput.match(/^(\d+)x?\s+/);

	if (m_amount)
	{
		const amount = parseInt(m_amount[1]);

		const name = rawInput.slice(m_amount[0].length).replace(_setRegex, '').replace(_collectorRegex, '').trim();

		const set = rawInput.match(_setRegex);
		const collectors = rawInput.match(_collectorRegex);

		return {
			name,
			amount,
			// set code will be in second matched group
			set: set ? set[2] : undefined,
			collectors: collectors ? parseInt(collectors[0]) : undefined,
		}
	}

	throw new SyntaxError(`Invalid format: ${rawInput}`);
}

/**
 * AMOUNT NAME (SET) COLLECTORS
 * @example 2 Trelasarra, Moon Dancer (AFR) 236
 */
export function toCardString(card: ICard)
{
	return [
		card.amount || 1,
		toCardStringWithoutAmount(card),
	].filter(s => s ?? false).join(' ')
}

/**
 * NAME (SET) COLLECTORS
 * @example Trelasarra, Moon Dancer (AFR) 236
 */
export function toCardStringWithoutAmount(card: ICardWithoutAmount)
{
	return [
		card.name,
		card.set?.length ? `(${card.set})` : undefined,
		card.collectors,
	].filter(s => s ?? false).join(' ')
}

export function toDeckListString(deck: IDeck)
{
	let output: string[] = [];

	if (deck.commander)
	{
		output.push(`Commander`);
		output.push(deck.commander.toCardString());
		output.push('');
	}

	if (deck.companion)
	{
		output.push(`Companion`);
		output.push(deck.companion.toCardString());
		output.push('');
	}

	if (deck.deck?.length)
	{
		output.push(`Deck`);
		deck.deck.forEach(card =>
		{
			output.push(card.toCardString());
		});
		output.push('');
	}

	if (deck.sideboard?.length)
	{
		output.push(`Sideboard`);
		deck.sideboard.forEach(card =>
		{
			output.push(card.toCardString());
		});
		output.push('');
	}

	return output.join('\n');
}
