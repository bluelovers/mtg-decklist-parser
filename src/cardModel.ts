import { parseString, toCardString } from './util';
import { ICard, ICardXmlObject } from './types';

export class CardModel implements ICard
{
	name: string;
	amount: number;
	set?: string;
	collectors?: number;
	mtgoID?: string;

	constructor(rawInput: string | ICardXmlObject | ICard)
	{
		const { amount, name, set, collectors, mtgoID } = typeof rawInput === 'string' ?
			this.parseString(rawInput) :
			this.parseObject(rawInput);

		this.name = name;
		this.amount = amount;
		this.set = set;
		this.collectors = collectors;
		this.mtgoID = mtgoID;
	}

	protected parseString(rawInput: string): ICard
	{
		return parseString(rawInput)
	}

	protected parseObject<T extends ICardXmlObject | ICard>(rawInputObject: T): ICard
	{
		return {
			// @ts-ignore
			name: rawInputObject.Name ?? rawInputObject.name,
			// @ts-ignore
			amount: parseInt(rawInputObject.Quantity ?? rawInputObject.amount),
			// @ts-ignore
			mtgoID: rawInputObject.CatID ?? rawInputObject.mtgoID,
			// @ts-ignore
			set: rawInputObject.set,
			// @ts-ignore
			collectors: rawInputObject.collectors,
		};
	}

	toCardString()
	{
		return toCardString(this)
	}

}
