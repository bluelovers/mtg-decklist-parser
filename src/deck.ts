import { CardModel } from './cardModel';

export class Deck
{
	readonly valid: boolean = false;
	deck: CardModel[] = [];
	sideboard: CardModel[] = [];
	companion?: CardModel = null;
	commander?: CardModel = null;
}
