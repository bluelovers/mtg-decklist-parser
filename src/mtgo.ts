import parser, { ValidationError } from 'fast-xml-parser';
import { CardModel } from './cardModel';
import { Deck } from './deck';

const _commanderAnnotation = '16777728';

export class MTGO extends Deck
{
	constructor(xml: string, logError = true)
	{
		super();

		xml = xml.trim();

		const output = parser.validate(xml);

		// @ts-ignore
		this.valid = output === true;

		if (this.valid)
		{
			const parsed = parser.parse(xml, {
				attributeNamePrefix: '',
				ignoreAttributes: false,
			}) as {
				Deck: {
					Cards: {
						Annotation?: typeof _commanderAnnotation,
						Sideboard?: 'true',
						Quantity: string,
						Name: string,
						CatID: string,
					}[]
				}
			};

			parsed.Deck.Cards.forEach(card =>
			{
				if (card.Annotation === _commanderAnnotation)
				{
					this.commander = new CardModel(card);
				}
				else if (card.Sideboard === 'true')
				{
					this.sideboard.push(new CardModel(card));
				}
				else
				{
					this.deck.push(new CardModel(card));
				}
			});
		}
		else if (logError)
		{
			console.error((output as ValidationError).err);
		}
	}
}
