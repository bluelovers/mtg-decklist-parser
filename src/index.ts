import { Decklist } from './deck/decklist';
import { MTGO } from './deck/mtgo';
import { MtgifyDecklist } from './deck/mtgify/index';

export { CardModel } from './cardModel';
export { parseString, toCardString, toDeckListString, toCardStringWithoutAmount } from './util';

export { Decklist, MTGO, MtgifyDecklist };

export function autoParse(rawInput: string | Uint8Array): MTGO | Decklist | MtgifyDecklist
{
	let deck = new MTGO(rawInput, false);

	if (deck.valid)
	{
		return deck
	}

	// @ts-ignore
	deck = new Decklist(rawInput);

	if (deck.valid)
	{
		return deck as any as Decklist
	}

	let deck3 = new MtgifyDecklist(rawInput);

	if (deck3.valid)
	{
		return deck3
	}

	return deck as any as Decklist;
}

export default autoParse
