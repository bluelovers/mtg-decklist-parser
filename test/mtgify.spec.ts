import { join, posix, relative } from 'path';
import FastGlob from '@bluelovers/fast-glob';
import { readFile } from 'fs-extra';
import autoParse from '../src/index';
import { __root } from './__root';
import { EnumDecklistType, SymDecklistType } from '../src/types';

describe(`deck:mtgify`, () =>
{

	FastGlob.sync<string>([
			'*.{dec,txt}',
		], {
			cwd: join(__root, 'test', 'deck', 'mtgify'),
			absolute: true,
		})
		.forEach(file =>
		{

			test(posix.normalize(relative(__root, file)), async () =>
			{

				let rawInput = await readFile(file)

				let actual = autoParse(rawInput);

				expect(actual).toMatchSnapshot({
					valid: true,
					[SymDecklistType]: EnumDecklistType.mtgify,
				});

			});

		})
	;

});
