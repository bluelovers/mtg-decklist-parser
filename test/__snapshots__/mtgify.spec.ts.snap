// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deck:mtgify test\\deck\\mtgify\\jodah.dec 1`] = `
Object {
  "commander": null,
  "companion": null,
  "deck": Array [
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Jodah, Archmage Eternal *CMDR*",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Arid Mesa",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Azorius Chancery",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Badlands",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Bayou",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Bloodstained Mire",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Bojuka Bog",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Boros Garrison",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Breeding Pool",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Dimir Aqueduct",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Flooded Strand",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Golgari Rot Farm",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Gruul Turf",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Halimar Depths",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Izzet Boilerworks",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Krosan Verge",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Marsh Flats",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Misty Rainforest",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Orzhov Basilica",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Overgrown Tomb",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Plateau",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Polluted Delta",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Rakdos Carnarium",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Savannah",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Scalding Tarn",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Scattered Groves",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Scrubland",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Selesnya Sanctuary",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Sheltered Thicket",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Simic Growth Chamber",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Skyline Cascade",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Stomping Ground",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Taiga",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Temple Garden",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Tropical Island",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Tundra",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Underground Sea",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Verdant Catacombs",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Volcanic Island",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Windswept Heath",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Wooded Foothills",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Balancing Act",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Bring to Light",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Broken Bond",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "By Force",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Decimate",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Decree of Pain",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Fabricate",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Farseek",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Hunting Wilds",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "In Garruk's Wake",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Kindred Dominance",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Life from the Loam",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Merciless Eviction",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Natural Balance",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Necromantic Selection",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Ranger's Path",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Razia's Purification",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Recurring Insight",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Skyshroud Claim",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Spoils of Victory",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Summer Bloom",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Tempt with Discovery",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Time Spiral",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Wargate",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Wheel of Fortune",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Whispering Madness",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Windfall",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Constant Mists",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Cyclonic Rift",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Fact or Fiction",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Frantic Search",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Realms Uncharted",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Sunder",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Garruk Wildspeaker",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Azusa, Lost but Seeking",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Budoka Gardener",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Courser of Kruphix",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Dread Cacodemon",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Kynaios and Tiro of Meletis",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Llanowar Scout",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Mina and Denn, Wildborn",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Muldrotha, the Gravetide",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Oracle of Mul Daya",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Sakura-Tribe Elder",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Skyshroud Ranger",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Tatyova, Benthic Druid",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Trinket Mage",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Walking Atlas",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Burgeoning",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Exploration",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Fall of the Thran",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Retreat to Coralhelm",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Rites of Flourishing",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Territorial Dispute",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Treachery",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Amulet of Vigor",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Candelabra of Tawnos",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Crucible of Worlds",
      "set": undefined,
    },
    CardModel {
      "amount": 1,
      "collectors": undefined,
      "mtgoID": undefined,
      "name": "Ward of Bones",
      "set": undefined,
    },
  ],
  "sideboard": Array [],
  "valid": true,
  Symbol(DecklistType): "mtgify",
}
`;
